File: Python/minimum_absolute_difference_in_bst.py
# https://leetcode.com/problems/minimum-absolute-difference-in-bst/description/
from utils.examination import Examiner, Testcase, Logger
from utils.tree import TreeNode


class Solution:
    def getMinimumDifference(self, root: TreeNode|None) -> int:
        min_diff: int = float('inf')
        if root is None:
            return min_diff

        biggest_node_in_left: TreeNode|None = root.left
        while biggest_node_in_left is not None and biggest_node_in_left.right is not None:
            biggest_node_in_left = biggest_node_in_left.right
        if biggest_node_in_left is not None:
            min_diff = min(min_diff, abs(biggest_node_in_left.val - root.val))
            if 1 == min_diff:
                return 1

        smallest_node_in_right: TreeNode|None = root.right
        while smallest_node_in_right is not None and smallest_node_in_right.left is not None:
            smallest_node_in_right = smallest_node_in_right.left
        if smallest_node_in_right is not None:
            min_diff = min(min_diff, abs(smallest_node_in_right.val - root.val))
            if 1 == min_diff:
                return 1

        return min(min_diff, self.getMinimumDifference(root.left), self.getMinimumDifference(root.right))


def examine()->None:
    Examiner(Solution(), 'getMinimumDifference').run((
        Testcase((TreeNode.from_list([236,104,701,None,227,None,911]),), 9),
        Testcase((TreeNode.from_list([4,2,6,1,3]),), 1),
        Testcase((TreeNode.from_list([1,0,48,None,None,12,49]),), 1),
    ))

File: TypeScript/solutions/minimum-absolute-difference-in-bst.ts
// https://leetcode.com/problems/minimum-absolute-difference-in-bst/description/
import { Examiner, Testcase, TreeNode } from '../utils';


function getMinimumDifference(root: TreeNode | null): number {
    let minDiff = Infinity;
    if (root === null) return minDiff;

    let biggestInLeft = root.left;
    while (biggestInLeft?.right !== null && biggestInLeft !== null) biggestInLeft = biggestInLeft.right;
    if (biggestInLeft !== null) {
        minDiff = Math.min(minDiff, Math.abs(biggestInLeft.val - root.val));
        if (minDiff === 1) return 1;
    }

    let smallestInRight = root.right;
    while (smallestInRight?.left !== null && smallestInRight !== null) smallestInRight = smallestInRight.left;
    if (smallestInRight !== null) {
        minDiff = Math.min(minDiff, Math.abs(smallestInRight.val - root.val));
        if (minDiff === 1) return 1;
    }

    return Math.min(minDiff, getMinimumDifference(root.left), getMinimumDifference(root.right));
}

export function examine(): void {
    new Examiner(getMinimumDifference).run([
        new Testcase([TreeNode.fromList([236, 104, 701, null, 227, null, 911])], 9),
        new Testcase([TreeNode.fromList([4, 2, 6, 1, 3])], 1),
        new Testcase([TreeNode.fromList([1, 0, 48, null, null, 12, 49])], 1),
    ]);
}
